### PR TITLE
[ncp] add vendor hook for property get/set handlers for vendor properties

### PR DESCRIPTION
**Pull request by abtink**

### CODE REVIEW

File: src/ncp/ncp_base.cpp [hunk 1]
```diff
@@ -1249,7 +1249,13 @@ otError NcpBase::HandleCommandPropertySet(uint8_t aHeader, spinel_prop_key_t aKe
     otError error = OT_ERROR_NONE;
     PropertyHandler handler = FindSetPropertyHandler(aKey);
 
-    if (handler == NULL)
+    if (handler != NULL)
+    {
+        mDisableStreamWrite = false;
+        error = (this->*handler)();
+        mDisableStreamWrite = true;
```
Comment (pvanhorn) [inline src/ncp/ncp_base.cpp]: do we want to disable stream write for vendor handlers as well?

Comment (abtink) [reply src/ncp/ncp_base.cpp]: I changed this in new pushed commit. We are actually enabling it so "set handler" can do logs/prints.

File: src/ncp/ncp_base.hpp [hunk 2]
```diff
@@ -709,6 +704,69 @@ class NcpBase
     void StopLegacy(void) { }
 #endif
 
+#if OPENTHREAD_ENABLE_NCP_VENDOR_HOOK
+    /**
+     * This method defines a vendor "command handler" hook to process vendor-specific spinel commands.
+     *
+     * @param[in] aHeader   The spinel frame header.
+     * @param[in] aHeader   The command key.
```
Comment (jwhui) [inline src/ncp/ncp_base.hpp]: typo: "aHeader" -> "aCommand"

File: src/ncp/ncp_base.hpp [hunk 3]
```diff
@@ -709,6 +704,69 @@ class NcpBase
...
+     * spinel frame is sent and removed from NCP buffer.
+     *
+     * (a) This can be used to track and verify that a vendor spinel frame response is delivered to the host (tracking
+     *     the frame using its tag).
+     *
+     * (b) It indicates that NCP buffer space is now available (since a spinel frame is removed). This can be used to
+     *     implement mechanisms to re-send a failed/pending response or an async spinel frame
+     *
+     */
+    void VendorHandleFrameRemovedFromNcpBuffer(NcpFrameBuffer::FrameTag aFrameTag);
```
Comment (jwhui) [inline src/ncp/ncp_base.hpp]: Document `aFrameTag`

File: src/ncp/example_vendor_hook.cpp [hunk 4]
```diff
@@ -52,6 +53,69 @@ otError NcpBase::VendorCommandHandler(uint8_t aHeader, unsigned int aCommand)
...
+otError VendorGetPropertyHandler(spinel_prop_key_t aPropKey)
+{
+    otError error = OT_ERROR_NONE;
+
+    switch (aPropKey)
+    {
+
+    // TODO: Implement your property get handlers here.
+    //
+    // Get handler should get/retrieve the property value and then encode and write the
```
Comment (jwhui) [inline src/ncp/example_vendor_hook.cpp]: nit: remove "get/"

File: src/ncp/example_vendor_hook.cpp [hunk 5]
```diff
@@ -52,6 +53,69 @@ otError NcpBase::VendorCommandHandler(uint8_t aHeader, unsigned int aCommand)
...
+otError VendorSetPropertyHandler(spinel_prop_key_t aPropKey)
+{
+    otError error = OT_ERROR_NONE;
+
+    switch (aPropKey)
+    {
+
+    // TODO: Implement your property set handlers here.
+    //
+    // Set handler  should first decode/parse the value from the input Spinel frame
```
Comment (jwhui) [inline src/ncp/example_vendor_hook.cpp]: nit: double space between "handler" and "should"
nit: choose either "decode" or "parse"

File: src/ncp/ncp_base.hpp [hunk 6]
```diff
@@ -709,6 +704,69 @@ class NcpBase
...
+
+    /**
+     * This method defines a vendor "get property handler" hook to process vendor spinel properties.
+     *
+     * The vendor handler should return `OT_ERROR_NOT_FOUND` status if it does not support "get" operation for the
+     * given property key. Otherwise, the vendor handler should behave like other property get handlers, i.e., it
+     * should get/retrieve the property value and then encode and write the value into the NCP buffer. If the "get"
```
Comment (jwhui) [inline src/ncp/ncp_base.hpp]: nit: remove "get/"

File: src/ncp/ncp_base.cpp [hunk 7]
```diff
@@ -1258,14 +1264,25 @@ otError NcpBase::HandleCommandPropertySet(uint8_t aHeader, spinel_prop_key_t aKe
...
+#if OPENTHREAD_ENABLE_NCP_VENDOR_HOOK
+        if (aKey >= SPINEL_PROP_VENDOR__BEGIN && aKey < SPINEL_PROP_VENDOR__END)
+        {
+            mDisableStreamWrite = false;
+            error = VendorSetPropertyHandler(aKey);
+            mDisableStreamWrite = true;
 
-    mDisableStreamWrite = false;
+            // An `OT_ERROR_NOT_FOUND` status from vendor handler indicates
+            // that it did not know/support the given property key. In that
```
Comment (jwhui) [inline src/ncp/ncp_base.cpp]: nit: "it did not know/support" -> "it does not support"

**Comment by abtink** [comment]
In new pushed commit, rebased on master to fix the conflict in `ncp_base.hpp`.

**Pull request merged**